src/server.js

```diff
@@ -0,0 +1,211 @@
+      RdHandler._reqObjTemplate.host = RdGlobalConfig.proxy.host;
+      RdHandler._reqObjTemplate.port = RdGlobalConfig.proxy.port;
+      
+      if (RdGlobalConfig.proxy.username && RdGlobalConfig.proxy.password) {
+        RdHandler._reqObjTemplate.headers['Proxy-Authorization'] = Utils.proxyAuthToBase64(RdGlobalConfig.proxy);
+      }
+
+      /**
+       * Sets the internal method to generate request options if external/upstream
+       * proxy exists
+       * @param {http.IncomingMessage} clientRequest 
+       * @returns {Object}
+       */
+      RdHandler._generateRequestOptions = function (clientRequest) {
+        var parsedClientUrl = url.parse(clientRequest.url);
+        var headersCopy = Object.assign({}, clientRequest.headers, RdHandler._reqObjTemplate.headers);
+        var requestOptions = Object.assign({}, RdHandler._reqObjTemplate);
+        requestOptions.path = parsedClientUrl.href;
+        requestOptions.method = clientRequest.method;
+        requestOptions.headers = headersCopy;
+        return requestOptions;
+      };
+    } else {
+
+      /**
+       * Sets the internal method to generate request options if external/upstream
+       * doesn't exists
+       * @param {http.IncomingMessage} clientRequest 
+       * @returns {Object}
+       */
+      RdHandler._generateRequestOptions = function (clientRequest) {
+        var parsedClientUrl = url.parse(clientRequest.url);
+        var requestOptions = Object.assign({}, RdHandler._reqObjTemplate);
+        requestOptions.host = parsedClientUrl.hostname;
+        requestOptions.port = parsedClientUrl.port || 80;
+        requestOptions.path = parsedClientUrl.path;
+        requestOptions.method = clientRequest.method;
+        requestOptions.headers = clientRequest.headers;
+        if (parsedClientUrl.auth) {
+          requestOptions.headers['authorization'] = Utils.proxyAuthToBase64(parsedClientUrl.auth);
+        }
+        return requestOptions;
+      };
+    }
+  },
+
+  /**
+   * Frames the error response based on the type of request.
+   * i.e., if its a request originating for Hub, the response
+   * is in the format which the client binding would understand.
+   * @param {Object} parsedRequest 
+   * @param {String} errorMessage 
+   */
+  _frameErrorResponse: function (parsedRequest, errorMessage) {
+    errorMessage += '. ' + constants.REQ_FAILED_MSG;
+    var parseSessionId = parsedRequest.path.match(/\/wd\/hub\/session\/([a-z0-9]+)\/*/);
+    if (parseSessionId) {
+      var sessionId = parseSessionId[1];
+      return {
+        data: {
+          sessionId: sessionId,
+          status: 13,
+          value: {
+            message: errorMessage,
+            error: constants.REQ_FAILED_MSG
+          },
+          state: 'error'
+        },
+        statusCode: 500
+      };
+    } else {
+      return {
+        data: {
+          message: errorMessage,
+          error: constants.REQ_FAILED_MSG
+        },
+        statusCode: 500
+      };
+    }
```

Comment:
What for `/wd/hub/session` for creating a new session?

Reply:
This is just to frame the response. As in, if the request is coming from a client binding, it shall be of the format `wd/hub/session`.. and thus if the request fails at the tool itself, i need to response in a manner which the client binding will understand.

Reply:
Cool 👍